packages/core/src/controllers/EnsController.ts

```diff
@@ -0,0 +1,161 @@
+import { NetworkController } from './NetworkController.js'
+import { NetworkUtil } from '@web3modal/common'
+import { EnsUtil } from '../utils/EnsUtil.js'
+
+const WC_NAME_SUFFIX = '.wcn.id'
+
+// -- Types --------------------------------------------- //
+type Suggestion = {
+  name: string
+  registered: boolean
+}
+
+export interface EnsControllerState {
+  suggestions: Suggestion[]
+  loading: boolean
+}
+
+type StateKey = keyof EnsControllerState
+
+// -- State --------------------------------------------- //
+const state = proxy<EnsControllerState>({
+  suggestions: [],
+  loading: false
+})
+
+// -- Controller ---------------------------------------- //
+export const EnsController = {
+  state,
+
+  subscribe(callback: (newState: EnsControllerState) => void) {
+    return sub(state, () => callback(state))
+  },
+
+  subscribeKey<K extends StateKey>(key: K, callback: (value: EnsControllerState[K]) => void) {
+    return subKey(state, key, callback)
+  },
+
+  async resolveName(name: string) {
+    try {
+      return await BlockchainApiController.lookupEnsName(name)
+    } catch (e) {
+      const error = e as BlockchainApiEnsError
+      throw new Error(error?.reasons?.[0]?.description || 'Error resolving name')
+    }
+  },
+
+  async isNameRegistered(name: string) {
+    try {
+      await BlockchainApiController.lookupEnsName(name)
+
+      return true
+    } catch {
+      return false
+    }
+  },
+
+  async getSuggestions(name: string) {
+    try {
+      state.loading = true
+      state.suggestions = []
+      const response = await BlockchainApiController.getEnsNameSuggestions(name)
+      state.suggestions =
+        response.suggestions.map(suggestion => ({
+          ...suggestion,
+          name: suggestion.name.replace(WC_NAME_SUFFIX, '')
+        })) || []
+      state.loading = false
+
+      return state.suggestions
+    } catch (e) {
+      state.loading = false
+      const errorMessage = this.parseEnsApiError(e, 'Error fetching name suggestions')
+      throw new Error(errorMessage)
+    }
+  },
+
+  async getNamesForAddress(address: string) {
+    try {
+      const network = NetworkController.state.caipNetwork
+      if (!network) {
+        return []
+      }
+
+      const response = await BlockchainApiController.reverseLookupEnsName({ address })
+
+      return response
+    } catch (e) {
+      const errorMessage = this.parseEnsApiError(e, 'Error fetching names for address')
+      throw new Error(errorMessage)
+    }
+  },
+
+  async registerName(name: string) {
+    const network = NetworkController.state.caipNetwork
+    if (!network) {
+      throw new Error('Network not found')
+    }
+    const address = AccountController.state.address
+    const emailConnector = ConnectorController.getAuthConnector()
+    if (!address || !emailConnector) {
+      throw new Error('Address or auth connector not found')
+    }
+    state.loading = true
+    const message = JSON.stringify({
```

Comment:
```suggestion
    const network = NetworkController.state.caipNetwork
    const emailConnector = ConnectorController.getAuthConnector()
    
    if (!network) {
      throw new Error('Network not found')
    }
    
     if (!address || !emailConnector) {
      throw new Error('Address or auth connector not found')
    }
    
    const address = AccountController.state.address
    state.loading = true
    
    const message = JSON.stringify({
```

Reply:
don't really understand the suggested change, this code will break since address is not defined 🤔

Reply:
This was just to make it look better, yeah forgot to add address line - sorry